### PR TITLE
Allow for custom samplers in dataloaders

### LINKED ISSUE #1995: Refactor AnnDataLoader to accept custom samplers
It looks like the way https://github.com/scverse/scvi-tools/blob/main/scvi/dataloaders/_ann_dataloader.py is written it'll always overwrite any custom samplers supplied to `**data_loader_kwargs` with a default torch sampler. For some models like CPA, it would help to have a sampler that takes into account cell labels since it implicitly assumes there are control cells in each batch. 

Would it be possible to only add the default sampler if the argument `sampler` isn't supplied to `**data_loader_kwargs`?

Thanks!
Yan

### PR DESCRIPTION
**Pull request by martinkim0**
Closes #1995

### CODE REVIEW

File: scvi/dataloaders/_ann_dataloader.py [hunk 1]
```diff
@@ -24,38 +26,58 @@ class AnnDataLoader(DataLoader):
...
     data_loader_kwargs
-        Keyword arguments for :class:`~torch.utils.data.DataLoader`
+        Keyword arguments for :class:`~torch.utils.data.DataLoader`.
     """
 
     def __init__(
         self,
-        adata_manager: AnnDataManager,
+        adata_manager: Optional[AnnDataManager] = None,
+        dataset: Optional[Dataset] = None,
```
Comment (adamgayoso) [inline scvi/dataloaders/_ann_dataloader.py:57]: let's not expose this

File: scvi/dataloaders/_ann_dataloader.py [hunk 2]
```diff
@@ -66,21 +88,32 @@ def __init__(
...
+            sampler_cls = SequentialSampler if not shuffle else RandomSampler
+            sampler = BatchSampler(
+                sampler=sampler_cls(self.dataset),
+                batch_size=batch_size,
+                drop_last=drop_last,
+            )
+        if isinstance(sampler, BatchSampler):
+            batch_size = None  # disables PyTorch automatic batching
+            shuffle = False
+            drop_last = False
```
Comment (adamgayoso) [inline scvi/dataloaders/_ann_dataloader.py:102]: what is going on here?

File: scvi/dataloaders/_ann_dataloader.py [hunk 3]
```diff
@@ -66,21 +88,32 @@ def __init__(
...
+
+        if sampler is None:
+            sampler_cls = SequentialSampler if not shuffle else RandomSampler
+            sampler = BatchSampler(
+                sampler=sampler_cls(self.dataset),
+                batch_size=batch_size,
+                drop_last=drop_last,
+            )
+        if isinstance(sampler, BatchSampler):
+            batch_size = None  # disables PyTorch automatic batching
```
Comment (adamgayoso) [inline scvi/dataloaders/_ann_dataloader.py:100]: this is important to add as a data loader kwarg when using the batchsampler

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/dataloaders/_ann_dataloader.py [hunk 4]
```diff
@@ -24,38 +26,58 @@ class AnnDataLoader(DataLoader):
...
+        `adata_manager` is not or if `sampler` is provided.
     shuffle
-        Whether the data should be shuffled
+        Whether the data should be shuffled.
     indices
-        The indices of the observations in the adata to load
+        The indices of the observations in the adata to load.
+    sampler
+        The sampler used for the dataloader. If `None`, a
+        :class:`~torch.utils.data.BatchSampler` will be used.
```
Comment (adamgayoso) [inline scvi/dataloaders/_ann_dataloader.py:40]:
```suggestion
        :class:`~torch.utils.data.BatchSampler` will be used.
        This enables access anndata with a minibatch of indices,
        and makes data loading much faster for sparse arrays.
```

File: scvi/dataloaders/_ann_dataloader.py (hunk 3 above)
Comment (adamgayoso) [reply scvi/dataloaders/_ann_dataloader.py:100]: it should be forced as a data loader kwarg under our default sampler

File: scvi/dataloaders/_ann_dataloader.py (hunk 2 above)
Comment (martinkim0) [reply scvi/dataloaders/_ann_dataloader.py:102]: We need to force these kwargs if the custom sampler they are passing in is a batch sampler, otherwise torch throws an error

**Comment by adamgayoso** [comment]
Close due to #2036
